addons/mass_mailing/controllers/main.py

```diff
@@ -167,23 +169,16 @@ def view(self, mailing_id, email=None, res_id=None, token=""):
             if not self._valid_unsubscribe_token(mailing_id, res_id, email, str(token)) and not request.env.user.has_group('mass_mailing.group_mass_mailing_user'):
                 raise exceptions.AccessDenied()
 
-            res = mailing.convert_links()
-            base_url = mailing.get_base_url().rstrip('/')
-            urls_to_replace = [
-                (base_url + '/unsubscribe_from_list', mailing._get_unsubscribe_url(email, res_id)),
-                (base_url + '/view', mailing._get_view_url(email, res_id))
-            ]
-            for url_to_replace, new_url in urls_to_replace:
-                if url_to_replace in res[mailing_id]:
-                    res[mailing_id] = res[mailing_id].replace(url_to_replace, new_url if new_url else '#')
-
-            res[mailing_id] = res[mailing_id].replace(
-                'class="o_snippet_view_in_browser"',
-                'class="o_snippet_view_in_browser" style="display: none;"'
-            )
+            html = str(request.env['mail.render.mixin'].sudo()._render_template_qweb(
+                mailing.body_html, mailing.mailing_model_real, [res_id], add_context=None, options={})[res_id])
+            # Update generic URLs (without parameters) to final ones
+            html = html.replace('/unsubscribe_from_list',
+                                urllib.parse.quote(mailing._get_unsubscribe_url(email, res_id), safe=':/?=&%'))
```

Comment:
Mmmh ? Seems complicated ?

Reply:
As we now keep a markupsafe from the start to end, the "replace" method of markupsafe takes care of the escaping so we don’t need it anymore.